### ISSUE TITLE
Printing modes

### ISSUE THREAD

**Issue by glebm** [vishesh/racketscript#92]
Perhaps the following functions should be on every primitive:

```typescript
toStringForWrite(): String // formerly toRawString
toStringForPrintAsExpression(): String // defaults to toStringForWrite
toString(): String // for display
```

Both should produce output readable by Racket (they mostly already do).
We could go even further if we decide to support Unicode property testing (then we can test e.g. if a character should be escaped or not).

They should return JS-native strings, as I think most of their use-cases will be IO.
We can later write optimized `toUStringFor*` versions if that proves necessary.

**Comment by vishesh** [comment]
I agree with renames and need to support printing modes properly. Unfortunately, it is not immediately important and I wont be able to properly research or even review that now. Feel free to research and write and implementation meanwhile.

This is likely to be bigger work than we know. Couple of things that I can think of right now is:  1. Circular data (I've some small WIP patch for this somewhere to check equality). 2. Output of write should be readable, which means we should be able to reconstruct the data. One easy way is to build a JSON string, which means we first need to convert our object to a typical JS object. This brings the question of doing an alternative representation such as using array/object with tags.

**Comment by glebm** [comment]
> One easy way is to build a JSON string

I propose using a format readable by Racket's `read`, not JSON.
For JSON, Racket already has `read-json` and `write-json`. https://docs.racket-lang.org/json/

**Comment by vishesh** [comment]
The JS object notation can be useful for FFI. Racket JSON library is not useful for  our case. `read` is definitely out of scope for now. Writing is much easier so no trouble even if we do that now. So if you plan to do it, you don't have to worry about (1) and (2) for now.

The interface itself IMO should correspond to functions in Racket instead of printing strings. 

```
write | display | printExpr : Port -> Void
```

**Comment by vishesh** [comment]
On side note, do keep performance and usability implications for writing Node or browser apps. In this case being compatible with Racket is not an absolute concern.

**Comment by glebm** [comment]
>  Racket JSON library is not useful for our case

If the data over the network is encoded as JSON, `read-json` should be used to read it.
If the data should be sent as JSON, `write-json` should be used to generate it.
RacketScript can provide optimized versions of these methods that are implemented using the platform-native `JSON` methods.

I'm confused why you think `read`/`write` should have anything to do with JSON and usability wrt Node/browser apps.

**Comment by vishesh** [comment]
I didn't not organize my thoughts clearly, and got side-tracked thinking about other issues around this area. I hope this wont be a big change, so it would be good to have this. The reason I'm avoiding spending too much time and effort on this for now is because currently for us `write`/`display` is pretty much only useful in fixture tests. Using them for debugging using Chrome tools is not great. And these are kinds the things that can easily drift away from Racket.

Regarding library, I meant it is not useful as we could just use JS functions instead to convert them to JS objects. The run-time should just care about producing and consuming JS objects to build things optimally.

> I'm confused why you think read/write should have anything to do with JSON and usability wrt Node/browser apps.

I was thinking if we can make use of Ports to model components like TextField or Sockets or whatever. We don't have `read` to read back those S-exp back, but JSON parsing is for free.  I did miss the part that JS is async - so that's one big difference for write.

**Comment by vishesh** [comment]
This is a related issue #23.

**Comment by glebm** [comment]
> The interface itself IMO should correspond to functions in Racket instead of printing strings.

Done.

 > currently `write/display` is pretty much only useful in fixture tests.

With a mostly functional `print/write/display`, we can get a REPL (in the console using nodejs or in browser).

**Comment by vishesh** [comment]
Resolved with PR #126. Thanks!
